proto-lens-benchmarks/src/Data/ProtoLens/BenchmarkUtil.hs

```diff
@@ -0,0 +1,77 @@
+-- Copyright 2016 Google Inc. All Rights Reserved.
+--
+-- Use of this source code is governed by a BSD-style
+-- license that can be found in the LICENSE file or at
+-- https://developers.google.com/open-source/licenses/bsd
+
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | A utility library for writing proto-lens benchmarks.
+module Data.ProtoLens.BenchmarkUtil (protoBenchmark, benchmarkMain) where
+
+import Control.DeepSeq (NFData)
+import Criterion.Main
+import qualified Criterion.Main.Options as Criterion
+import qualified Data.ByteString as BS (length)
+import Data.Maybe (fromMaybe)
+import Data.ProtoLens
+import Options.Applicative
+
+-- | Generate a group of benchmarks for encoding and decoding the given proto
+-- message. Includes benchmarks for decoding to both weak head normal form and
+-- normal form (i.e. all fields fully forced.) There is only one encoding
+-- benchmark because protos are encoded as strict ByteStrings, so there is no
+-- difference between normal form and weak head normal form.
+protoBenchmark
+    :: forall a.
+       (NFData a, Message a)
+    => String
+    -- ^ Benchmark name.
+    -> a
+    -- ^ Protocol buffer message to encode/decode.
+    -> Benchmark
+protoBenchmark groupName proto =
+    bgroup
+        fullGroupName
+        [ bench "encode" $ nf encodeMessage proto
+        , bgroup
+              "decode"
+              [ bench "whnf" $ whnf decodeMessage' encodedProto
+              , bench "nf" $ nf decodeMessage' encodedProto]]
+  where
+    -- We must indicate to the compiler that we want to decode to the same
+    -- message type as the input proto.
+    decodeMessage' bytes = (decodeMessage bytes :: Either String a)
```

Comment:
Should we use `decodeMessageOrDie` instead of `decodeMessage`, so that if there's a bug in the library we'd catch it rather than reporting incorrect benchmark timings?  (I'm not sure how criterion handles exceptions.)